core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java

```diff
@@ -212,17 +212,23 @@ public URIStatus getStatus(AlluxioURI path) {
 
   @Override
   public URIStatus getStatus(AlluxioURI path, final GetStatusPOptions options) {
-    return callWithReturn(() -> transformStatus(mUfs.get().getStatus(path.getPath())));
+    String ufsFullPath = PathUtils.concatPath(mRootUFS,
+        CommonUtils.stripPrefixIfPresent(path.getPath(), mRootUFS.getPath()));
```

Comment:
Should we first test if the `path` URI sent by client is within the same root UFS of this `UfsBaseFileSystem`? I don't think it makes sense to try to access something that is already known to be outside of the root UFS. 

If `mRootUFS` is `hdfs://namenode/` and `path` is `s3://bucket/file`, it makes no sense to try to get the status of `hdfs://namenode/file`. Also, what this code does here is creating something like `hdfs://namenode/s3://bucket/file` which is invalid URI.

Reply:
In Dora, we will only have a single UFS defined by:
`alluxio.dora.client.ufs.root`
We will not have multiple mount tables.

So, the `path` must be the Alluxio internal path.
It will not, should not contain any Provider (s3://, pdfs://, ...) information.

Reply:
I know, but `UfsBaseFileSystem` is not only meant for Dora. It should be robust enough to handle that even outside the context of Dora.

Reply:
The `path` is Alluxio Internal Path.
It should be joined with the "ufs root" before passing to ufs.
Otherwise the UFS will not recognize this path, and treat it as invalid.

Reply:
Yes, exactly, so we should do a precondition check and reject anything that is neither an Alluxio path (with no explicit scheme and authority), nor starts with the root UFS URI. Simply calling `CommonUtils.stripPrefixIfPresent(path.getPath(), mRootUFS.getPath())` cannot do this.
For reference, see [`alluxio.client.file.BaseFileSystem#checkUri`](https://github.com/Alluxio/alluxio/blob/1c781f7de19fbcdeddb1d15c8866bb0735f4c520/core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java#L523)

Reply:
`alluxio.client.file.BaseFileSystem#checkUri` is in another class, and we can not call it here.

The input, `ufsStatus` is not from user input. It is from getStatus() and listStatus() from this class.
The format of this input should be considered as valid?

We can add more checking here to validate the path scheme if needed.
@LuQQiu

Reply:
Do you like the following patch, @dbw9580 ?

```
diff --git a/core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java b/core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java
index 7bf66e26d6..ff322afaed 100644
--- a/core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java
+++ b/core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java
@@ -12,6 +12,7 @@
 package alluxio.client.file.ufs;
 
 import alluxio.AlluxioURI;
+import alluxio.client.file.BaseFileSystem;
 import alluxio.client.file.FileInStream;
 import alluxio.client.file.FileOutStream;
 import alluxio.client.file.FileSystem;
@@ -80,7 +81,7 @@ import javax.annotation.concurrent.ThreadSafe;
  * UFS. No Alluxio server is involved.
  */
 @ThreadSafe
-public class UfsBaseFileSystem implements FileSystem {
+public class UfsBaseFileSystem extends BaseFileSystem implements FileSystem {
   private static final Logger LOG = LoggerFactory.getLogger(UfsBaseFileSystem.class);
 
   /** Used to manage closeable resources. */
@@ -97,6 +98,7 @@ public class UfsBaseFileSystem implements FileSystem {
    * @param options the ufs file system options
    */
   public UfsBaseFileSystem(FileSystemContext fsContext, UfsFileSystemOptions options) {
+    super(fsContext);
     Preconditions.checkNotNull(fsContext);
     Preconditions.checkNotNull(options);
     mFsContext = fsContext;
@@ -408,6 +410,9 @@ public class UfsBaseFileSystem implements FileSystem {
    */
   private URIStatus transformStatus(UfsStatus ufsStatus) {
     String path = CommonUtils.stripPrefixIfPresent(ufsStatus.getName(), mRootUFS.toString());
+    // adding the leading "/" to the beginning of this path.
+    path = PathUtils.concatPath("/", path);
+    super.checkUri(new AlluxioURI(path));
     AlluxioURI ufsUri = new AlluxioURI(PathUtils.concatPath(mRootUFS, path));
     FileInfo info = new FileInfo().setName(ufsUri.getName())
         .setPath(path)
```


we can have ufsBaseFileSystem to extend BaseFileSystem and re-use the checkUri() method.

Reply:
I'm not sure about extending from `BaseFileSystem`. Can we just copy paste that piece of code for `checkUri`?

Reply:
We can add a method like `checkUri(AlluxioURI path)` here to check whether the path string contains schema or not.

Reply:
Copy & paste is not the right way to __duplicate__ code.
Maybe I can move this function to some common place, e.g. in AlluxioUri.checkUri().
What do you think?

Reply:
But again, as I said, this path is purely internal and returned from listStatus() or getStatus().
It should be a valid ufs path. No need to check it again.

Reply:
> Can we just copy paste that piece of code for checkUri?

Sorry about not making this clear. Apparently the implementation in `BaseFileSystem` cannot be reused as is, since it is trying to acquire the master authority to check if the authorities match. We don't need that for a `UfsBaseFileSystem` since no master should be involved at all. The implementation of `checkUri` for `UfsBaseFileSystem` should be as simple as

```java
if (!mRootUFS.isAncestorOf(path)) {
  throw new InvalidPathException(...);
}
```

Reply:
OK. Let me try.